File: proto/test/Test/Ledger.hs
{-# LANGUAGE OverloadedStrings #-}
-- | The cross-plan ledger: reservation at request, refusal of overlap and of
-- a held exclusivity class, rehearsals reserving nothing, release on close.
module Test.Ledger (tests) where

import Data.Text (Text)
import Rue.Proto.Interference (Fact (..))
import Rue.Proto.Ledger
import Test.Tasty (TestTree, testGroup)
import Test.Tasty.HUnit (assertBool, assertFailure, testCase, (@?=))

inst :: Text -> [Fact] -> Instance
inst i fs = Instance {instId = i, instHost = "db-01", instUmbra = fs, instExclusivity = Nothing, instRehearsal = False}

-- | Reserve one instance in an empty ledger, or fail the test.
reserved :: Instance -> IO Ledger
reserved i = either (\(c, why) -> assertFailure ("could not reserve " <> show c <> ": " <> show why)) pure (request i empty)

pf :: Fact
pf = Fact "file:/etc/pf.conf" Nothing

keysA, keysB :: Fact
keysA = Fact "file:/root/.ssh/authorized_keys" (Just "rue-a")
keysB = Fact "file:/root/.ssh/authorized_keys" (Just "rue-b")

tests :: TestTree
tests =
  testGroup
    "ledger"
    [ testCase "a second plan overlapping a pending umbra is refused at request with R0203" $
        case request (inst "a" [pf]) empty >>= request (inst "b" [pf]) of
          Left (code, _) -> code @?= R0203
          Right _ -> assertBool "admitted an overlap" False
    , testCase "distinct regions on one fact coexist" $
        assertBool "refused disjoint regions" (either (const False) (const True) (request (inst "a" [keysA]) empty >>= request (inst "b" [keysB])))
    , testCase "a held exclusivity class refuses with R0101 regardless of umbra" $
        case request (inst "a" []) {instExclusivity = Just "corpse-1"} empty >>= request (inst "b" [pf]) {instExclusivity = Just "corpse-1"} of
          Left (code, _) -> code @?= R0101
          Right _ -> assertBool "admitted a class collision" False
    , testCase "a rehearsal reserves nothing and is never blocked" $ do
        l <- reserved (inst "a" [pf])
        case request (inst "r" [pf]) {instRehearsal = True} l of
          Right l' -> length (holdings l') @?= 1
          Left _ -> assertBool "rehearsal blocked" False
    , testCase "release admits what was refused" $ do
        l <- reserved (inst "a" [pf])
        assertBool "still refused after release" (either (const False) (const True) (request (inst "b" [pf]) (release "a" l)))
    , testCase "different hosts never interfere" $
        assertBool "cross-host refusal" (either (const False) (const True) (request (inst "a" [pf]) empty >>= request (inst "b" [pf]) {instHost = "db-02"}))
    ]

File: core/tests/ledger.rs
//! The cross-plan ledger: reservation at request, refusal of overlap and of a
//! held exclusivity class, rehearsals reserving nothing, release on close.

use rue_core::interference::Fact;
use rue_core::ledger::*;

fn inst(id: &str, umbra: Vec<Fact>) -> Instance {
    Instance {
        id: id.into(),
        host: "db-01".into(),
        umbra,
        exclusivity: None,
        rehearsal: false,
    }
}

fn pf() -> Fact {
    Fact::new("file:/etc/pf.conf", None)
}

#[test]
fn a_second_plan_overlapping_a_pending_umbra_is_refused_with_r0203() {
    let l = Ledger::new().request(inst("a", vec![pf()])).unwrap();
    assert_eq!(
        l.request(inst("b", vec![pf()])).map(|_| ()).unwrap_err().0,
        LedgerCode::R0203
    );
}

#[test]
fn distinct_regions_on_one_fact_coexist_and_the_same_anchor_never_does() {
    let a = Fact::new("file:/root/.ssh/authorized_keys", Some("rue-a"));
    let b = Fact::new("file:/root/.ssh/authorized_keys", Some("rue-b"));
    let l = Ledger::new().request(inst("a", vec![a.clone()])).unwrap();
    assert!(l.request(inst("b", vec![b])).is_ok());
    // Two plans holding one anchor on one fact are the same umbra, not
    // two disjoint ones (5.2): the second is refused.
    assert_eq!(
        l.request(inst("c", vec![a])).map(|_| ()).unwrap_err().0,
        LedgerCode::R0203
    );
}

#[test]
fn a_held_exclusivity_class_refuses_with_r0101_regardless_of_umbra() {
    let l = Ledger::new()
        .request(Instance {
            exclusivity: Some("corpse-1".into()),
            ..inst("a", vec![])
        })
        .unwrap();
    assert_eq!(
        l.request(Instance {
            exclusivity: Some("corpse-1".into()),
            ..inst("b", vec![pf()])
        })
        .map(|_| ())
        .unwrap_err()
        .0,
        LedgerCode::R0101
    );
}

#[test]
fn a_rehearsal_reserves_nothing_and_is_never_blocked() {
    let l = Ledger::new().request(inst("a", vec![pf()])).unwrap();
    let l2 = l
        .request(Instance {
            rehearsal: true,
            ..inst("r", vec![pf()])
        })
        .unwrap();
    assert_eq!(l2.holdings().len(), 1);
}

#[test]
fn release_admits_what_was_refused() {
    let l = Ledger::new().request(inst("a", vec![pf()])).unwrap();
    assert!(l.release("a").request(inst("b", vec![pf()])).is_ok());
}

#[test]
fn different_hosts_never_interfere() {
    let l = Ledger::new().request(inst("a", vec![pf()])).unwrap();
    assert!(l
        .request(Instance {
            host: "db-02".into(),
            ..inst("b", vec![pf()])
        })
        .is_ok());
}
